examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2218 @@
+  // possible but none of them have much influence in theory or practice.
+  // We remind the reader that <code>CopyData</code> includes the class member
+  // <code>local_boundary_normal_map</code> in order to store these local
+  // contributions for the boundary map.
+
+  template <int dim>
+  void OfflineData<dim>::assemble()
+  {
+    lumped_mass_matrix = 0.;
+    norm_matrix        = 0.;
+    for (auto &matrix : cij_matrix)
+      matrix = 0.;
+    for (auto &matrix : nij_matrix)
+      matrix = 0.;
+
+    const unsigned int dofs_per_cell =
+      discretization->finite_element.dofs_per_cell;
+    const unsigned int n_q_points = discretization->quadrature.size();
+
+    /* This is the implementation of the scratch data required by WorkStream */
+    MeshWorker::ScratchData<dim> scratch_data(
+      discretization->mapping,
+      discretization->finite_element,
+      discretization->quadrature,
+      update_values | update_gradients | update_quadrature_points |
+        update_JxW_values,
+      discretization->face_quadrature,
+      update_normal_vectors | update_values | update_JxW_values);
+
+    {
+      TimerOutput::Scope t(
+        computing_timer,
+        "offline_data - assemble lumped mass matrix, and c_ij");
+
+      /* This is the implementation of the "worker" required by WorkStream */
+      const auto local_assemble_system = [&](const auto &cell,
+                                             auto &      scratch,
+                                             auto &      copy) {
+        auto &is_artificial             = copy.is_artificial;
+        auto &local_dof_indices         = copy.local_dof_indices;
+        auto &local_boundary_normal_map = copy.local_boundary_normal_map;
+        auto &cell_lumped_mass_matrix   = copy.cell_lumped_mass_matrix;
+        auto &cell_cij_matrix           = copy.cell_cij_matrix;
+
+        is_artificial = cell->is_artificial();
+        if (is_artificial)
+          return;
+
+        local_boundary_normal_map.clear();
+        cell_lumped_mass_matrix.reinit(dofs_per_cell, dofs_per_cell);
+        for (auto &matrix : cell_cij_matrix)
+          matrix.reinit(dofs_per_cell, dofs_per_cell);
+
+        const auto &fe_values = scratch.reinit(cell);
+
+        local_dof_indices.resize(dofs_per_cell);
+        cell->get_dof_indices(local_dof_indices);
+
+        std::transform(local_dof_indices.begin(),
+                       local_dof_indices.end(),
+                       local_dof_indices.begin(),
+                       [&](auto index) {
+                         return partitioner->global_to_local(index);
+                       });
+
+        /* We compute the local contributions for the  lumped mass
+         matrix entries m_i and and vectors c_ij */
+        for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+          {
+            const auto JxW = fe_values.JxW(q_point);
+
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
```

Comment:
we typically do q,i,j in that order.

Reply:
I would like to retain the current order so that the computation of the gradient doesn't end up in the innermost loop (not that it matters too much).